include/deal.II/base/mpi.h

```diff
@@ -708,6 +709,416 @@ namespace Utilities
+       * @param comm MPI Communicator
+       */
+      ConsensusAlgorithm_NBX(ConsensusAlgorithmProcess<T1, T2> &process,
+                             const MPI_Comm &                   comm);
+
+      /**
+       * Destructor.
+       */
+      virtual ~ConsensusAlgorithm_NBX();
+
+      /**
+       * Run consensus algorithm.
+       */
+      void
+      run() override;
+
+    private:
+#ifdef DEAL_II_WITH_MPI
+      std::vector<unsigned int> targets;
+
+      // for sending request
+      std::vector<std::vector<T1>> send_buffers;
+      std::vector<MPI_Request>     send_requests;
+
+      // for receiving answer to the request
+      std::vector<MPI_Request>     recv_requests;
+      std::vector<std::vector<T2>> recv_buffers;
+
+      // for sending answers to the requests
+      std::vector<std::vector<T2>>              request_buffers;
+      std::vector<std::shared_ptr<MPI_Request>> request_requests;
+
+      // request for barrier
+      MPI_Request barrier_request;
+#endif
+
+#ifdef DEBUG
+      std::set<unsigned int> requesting_processes;
+#endif
+
+      /**
+       * Check if all request answers have been received by this rank.
+       */
+      bool
+      check_own_state();
+
+      /**
+       * Signal to all other ranks that this rank has received all request
+       * answers via entering IBarrier.
+       */
+      void
+      signal_finish();
+
+      /**
+       * Check if all ranks have received all their request answers, i.e.
+       * all ranks have reached the IBarrier.
+       */
+      bool
+      check_global_state();
+
+      /**
+       * A request message from another rank has been received: process the
+       * request and send an answer.
+       */
+      void
+      process_requests();
+
+      /**
+       * Start to send all requests via ISend and post IRecvs for the incoming
+       * answer messages.
+       */
+      void
+      start_communication();
+
+      /**
+       * After all rank has received all answers, the MPI data structures can be
+       * freed and the received answers can be processed.
+       */
+      void
+      clean_up_and_end_communication();
+    };
+
+    /**
+     * This class implements ConsensusAlgorithm, using a two step approach. In
+     * the first step the source ranks are determined and in the second step
+     * a static sparse data exchange is performed.
+     *
+     * @note In contrast to ConsensusAlgorithm_NBX, this class splits the same
+     *       task into two distinct steps. In the first step, all processes are
+     *       identified who wants to send a request to this process. In the
+     *       second step, the data is exchanged. However, since - in the second
+     *       step - now it is clear how many requests have to be answered, i.e.
+     *       when this process can stop waiting for requests, no IBarrier is
+     *       needed.
+     *
+     * @note The function compute_point_to_point_communication_pattern is used
```

Comment:
```suggestion
     * @note The function compute_point_to_point_communication_pattern() is used
```
This way, doxygen automatically links to the documentation of that function.